dotnet/src/webdriver/DevTools/DevToolsSession.cs

```diff
@@ -437,116 +413,38 @@ private void OnTargetDetached(object sender, TargetDetachedEventArgs e)
+                await this.connection.Stop();
+                await this.ShutdownMessageQueue();
             }
         }
 
-        private async Task ReceiveMessage()
+        private async Task ShutdownMessageQueue()
         {
-            var cancellationToken = this.receiveCancellationToken.Token;
-            try
+            // Attempt to wait for the channel to empty before marking the
+            // writer as complete and waiting for the monitor task to end.
+            while (this.messageQueue.Reader.TryPeek(out _))
             {
-                var buffer = WebSocket.CreateClientBuffer(1024, 1024);
-                while (this.sessionSocket.State != WebSocketState.Closed && !cancellationToken.IsCancellationRequested)
-                {
-                    WebSocketReceiveResult result = await this.sessionSocket.ReceiveAsync(buffer, cancellationToken);
-                    if (!cancellationToken.IsCancellationRequested)
-                    {
-                        if (result.MessageType == WebSocketMessageType.Close && this.sessionSocket.State == WebSocketState.CloseReceived)
-                        {
-                            LogTrace("Got WebSocket close message from browser");
-                            await this.sessionSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, string.Empty, cancellationToken);
-                        }
-                    }
-
-                    if (this.sessionSocket.State == WebSocketState.Open && result.MessageType != WebSocketMessageType.Close)
-                    {
-                        using (var stream = new MemoryStream())
-                        {
-                            stream.Write(buffer.Array, 0, result.Count);
-                            while (!result.EndOfMessage)
-                            {
-                                result = await this.sessionSocket.ReceiveAsync(buffer, cancellationToken);
-                                stream.Write(buffer.Array, 0, result.Count);
-                            }
-
-                            stream.Seek(0, SeekOrigin.Begin);
-                            using (var reader = new StreamReader(stream, Encoding.UTF8))
-                            {
-                                string message = reader.ReadToEnd();
-
-                                // fire and forget
-                                // TODO: we need implement some kind of queue
-                                Task.Run(() => ProcessIncomingMessage(message));
-                            }
-                        }
-                    }
-                }
-            }
-            catch (OperationCanceledException)
-            {
-            }
-            catch (WebSocketException)
-            {
-            }
-            finally
-            {
-                this.sessionSocket.Dispose();
-                this.sessionSocket = null;
+                await Task.Delay(TimeSpan.FromMilliseconds(10));
```

Comment:
Seems algorithm is contiguously pinging something instead being event based.

Reply:
@nvborisenko Polling for a channel reader to be empty is the only way to ensure that no messages sent before the WebSocket connection was closed are lost. Since the socket connection is always closed before this method is called, the channel reader is guaranteed to be empty at some point. The comment before the loop attempts to explain this but is unclear about it. I'll update the comment.